File: src/go/helpers/search/binary.go
package search

import "fmt"

func BinarySearch(arr []int, target int, enableLogging bool) (int, int) {
	left := 0
	right := len(arr) - 1
	steps := 0

	for left <= right {
		steps++
		mid := (left + right) / 2

		if enableLogging {
			fmt.Printf("  Step %d: Check arr[%d] = %d (range: %d-%d)\n", steps, mid, arr[mid], left, right)
		}

		if arr[mid] == target {
			return mid, steps
		} else if arr[mid] < target {
			left = mid + 1
		} else {
			right = mid - 1
		}
	}

	return -1, steps
}

File: src/typescript/helpers/search/binary.ts
import { SearchResult } from "./types";

export function binarySearch(
  arr: number[],
  target: number,
  { loggingEnabled = false } = {}
): SearchResult {
  let steps = 0;
  let left = 0;
  let right = arr.length - 1;

  while (left <= right) {
    steps++;
    const mid = Math.floor((left + right) / 2);

    if (loggingEnabled) {
      console.log(
        `Step ${steps}: Check arr[${mid}] = ${arr[mid]} (range: ${left}-${right})`
      );
    }

    if (arr[mid] === target) {
      return { index: mid, steps };
    }

    if (arr[mid] < target) {
      left = mid + 1;
    } else {
      right = mid - 1;
    }
  }

  return { index: -1, steps };
}
